### PR TITLE
feat(SSP-2648): allow using request body for GET requests

### PR DESCRIPTION
**Pull request by volosh1n-EP**

### CODE REVIEW

File: spec/lib/remote_resource/request_spec.rb [hunk 1]
```diff
@@ -398,6 +398,26 @@ class RequestDummyWithCollectionPrefix < RequestDummy
...
+        end
+
+        it 'returns nil' do
+          expect(request.query).to be_nil
+        end
+      end
+
+      context "when connection_options[:params][:use_body] is present" do
+        let(:connection_options) do
+          { root_element: :data, params: { use_body: true, pseudonym: 'pseudonym', labels: [1, '2', 'three'], pagination: { page: 5, limit: 15, ordered: true } } }
```
Comment (murb) [inline spec/lib/remote_resource/request_spec.rb:414]: I think I don't like supporting `use_body` as a param tbh; we should settle on a single approach

Comment (murb) [reply spec/lib/remote_resource/request_spec.rb:414]: or is there a reason for wanting to support both?

Comment (volosh1n-EP) [reply spec/lib/remote_resource/request_spec.rb:414]: That's just to cover both cases in specs; code in lib folder wasn't modified

Comment (murb) [reply spec/lib/remote_resource/request_spec.rb:414]: got it, but to add it to the gem, we should default to a single approach, I suggest removing the option to have use_body as a param.

Comment (p8) [reply spec/lib/remote_resource/request_spec.rb:414]: `use_body` is a bit unclear as well. Maybe something like: `params_from_body`, `use_params_in_body`, `read_body_for_params` or `body_params`?

Comment (volosh1n-EP) [reply spec/lib/remote_resource/request_spec.rb:414]: @p8 we can do that, but use_body is already in many places in SSP and CM, we will need to go though all those places too then

**Comment by p8** [comment]
I can see the use case of this change.

Sending the request body with a GET is generally not recommend by the HTTP specifications.
The following article mentions some more reasons not to support it:
https://www.baeldung.com/cs/http-get-with-body

There is a [draft spec](https://httpwg.org/http-extensions/draft-ietf-httpbis-safe-method-w-body.html) for adding the QUERY HTTP method to the HTTP spec. I'm not sure if it would be possible to get that working but I'd prefer supporting an implementation that is more inline with the HTTP spec

**Comment by volosh1n-EP** [comment]
@p8 it's already works in SSP and CM, but with patching RemoteResource inside SSP and CM repo. The idea was to move that patch from repos to gem itself

File: spec/lib/remote_resource/request_spec.rb (hunk 1 above)
Comment (murb) [reply spec/lib/remote_resource/request_spec.rb]: @volosh1n-EP , that is the trade off of merging this to something reusable.

Perhaps `force_send_params_in_body`? (it is an instruction that won't / shouldn't reach the server)

**Comment by p8** [comment]
@volosh1n-EP I understand it would take resources to make changes.

If we are going to implement this in this gem other apps could start using it as well.
I'd like to be sure we have a implementation that is future proof and adheres to the HTTP standards as much as possible.

**Comment by murb** [comment]
> A client SHOULD NOT generate content in a GET request **unless it is made directly to an origin server that has previously indicated, in or out of band, that such a request has a purpose and will be adequately supported**

I think we satisfy the last part, also it is not the default.

**Comment by murb** [comment]
@p8 I'd add my approval flag to current situation if I could :) If you agree, please merge.

**Comment by p8** [comment]
> @p8, as to your comment about whether params in a body is a good idea...
> 
> > A client SHOULD NOT generate content in a GET request **unless it is made directly to an origin server that has previously indicated, in or out of band, that such a request has a purpose and will be adequately supported**
> 
> I think we satisfy the last part, also it is not the default.

We do for now, but this doesn't seem to be enforced.
Someone might start implementing this for a mobile endpoint and a proxy server could start caching responses based on the URL, ignoring the body.
Or someone might implement it for a third party API.

It's also unclear what happens if URL params conflict with body params. Which ones have precedence?
 
> As to his alternatives:
> 
>     * POST ... nogo, conflicts with create
> 
>     * QUERY ... nogo, not supported
> 
>     * SEARCH ... why advocate for SEARCH when that is not officially supported, hence the same argument can be made as params in GET?

I'm not advocating for `POST` or `SEARCH`, although a `POST` to a non-index action might work as well.
It's a bit unclear how much longer it will take for `QUERY` to be official. There is[ a lot of activity](https://github.com/httpwg/http-extensions/blob/main/draft-ietf-httpbis-safe-method-w-body.xml) on the draft proposal.

> Nice thing about this solution in the GET body is that it works completely transparent (at least with Rails (perhaps because of Rack)) on the server side.

Yes, I'm not sure if `QUERY` would even work in Rack of Rails.
You'd probably have to monkey patch Rack to support passing `_method=query` for GET requests (or _method=get for POST requests).
https://github.com/rack/rack/blob/df241355a7f122dc22437398267c1d5f0b27e1ad/test/spec_method_override.rb#L39-L54

Another alternative to the `use_body` parameter could be setting the `Content-Type` of the GET request to `application/x-www-form-urlencoded`.
This is also how `POST` works and how `QUERY` is drafted to work (which would make migrating easier if support for that is released).

**Comment by murb** [comment]
@p8 I think by making this more explicit, as now implemented using the `force_get_params_in_body`-option, it is clear that it is an exception. Note that DDY Remote Resource *is* the client, not the server. Server side no changes are needed to support this.

File: lib/remote_resource/request.rb [hunk 2]
```diff
@@ -80,6 +80,8 @@ def body
       @body ||= begin
         if [:put, :patch, :post].include?(http_action)
           attributes.to_json
+        elsif http_action == :get && connection_options[:force_get_params_in_body]
+          connection_options[:params].to_json
         else
           nil
         end
```
Comment (p8) [inline lib/remote_resource/request.rb:87]: A case statement would be bit more readable here I think:
```suggestion
        case http_action
        when :put, :patch, :post
          attributes.to_json
        when :get
          connection_options[:params].to_json if connection_options[:force_get_params_in_body]
        else
          nil
        end
```

**Comment by volosh1n-EP** [comment]
So, is it fine?

**Pull request merged**